File: warrior/src/main/java/com/dumbdogdiner/warrior/managers/ArenaManager.java
package com.dumbdogdiner.warrior.managers;

import com.dumbdogdiner.warrior.WithWarriorPlugin;
import com.dumbdogdiner.warrior.api.arena.Arena;
import com.dumbdogdiner.warrior.api.managers.WarriorArenaManager;
import com.dumbdogdiner.warrior.api.sessions.ArenaSession;
import com.dumbdogdiner.warrior.api.sessions.Session;
import com.dumbdogdiner.warrior.api.sessions.SessionType;
import com.dumbdogdiner.warrior.api.translation.Placeholders;
import com.dumbdogdiner.warrior.user.User;
import com.dumbdogdiner.warrior.api.util.JSONUtil;
import com.dumbdogdiner.warrior.utils.DefaultMessages;
import org.bukkit.entity.Player;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class ArenaManager implements WithWarriorPlugin, WarriorArenaManager {

    private final HashMap<Integer, Arena> arenaMap = new HashMap<>();

    public void registerArena(Arena arena) {
        arenaMap.put(arenaMap.size(), arena);
    }

    public List<Arena> getArenas() {
        return new ArrayList<>(arenaMap.values());
    }

    public Arena get(String arenaName) {
        for(Arena a : arenaMap.values()) {
            if(a.getName().equalsIgnoreCase(arenaName)) return a;
        }

        return null;
    }

    public void loadArenas() {
        File dataFolder = new File(JSONUtil.ARENA_DATA_PATH);
        File[] files = dataFolder.listFiles();

        if(files == null) return;

        int i = 0;
        for(File f : files) {
            registerArena(new Arena(f, i));
            i++;
        }

        int finalI = i;
        String msg = Placeholders.applyPlaceholders(DefaultMessages.LOADED_OBJECT, new HashMap<>() {
            {
                put("AMOUNT", Integer.toString(finalI));
                put("TYPE", "arena");
            }
        });
        getLogger().info(msg);
    }

    public List<User> getPlayers(Arena a) {
        return PlayerManager.getListOf(user -> {
            Session s = user.getSession();
            if(s.getType().equals(SessionType.GAME))
                return ((ArenaSession)s).getArena().equals(a);
            return false;
        });
    }

    public boolean isPlaying(Player player) {
        User user = PlayerManager.get(player.getUniqueId());
        if(user.getSession() == null) return false;
        return user.getSession().getType() == SessionType.GAME;
    }

    public ArenaSession getSession(Player player) {
        if(isPlaying(player)) {
            User user = PlayerManager.get(player.getUniqueId());
            return (ArenaSession) user.getSession();
        }

        return null;
    }

    public void remove(Arena a) {
        arenaMap.remove(a.getId());
        JSONUtil.removeFile(a);
    }
}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/managers/ArenaManager.kt
package com.dumbdogdiner.warrior.managers

import com.dumbdogdiner.warrior.WithWarriorPlugin
import com.dumbdogdiner.warrior.api.managers.WarriorArenaManager
import java.util.HashMap
import com.dumbdogdiner.warrior.api.arena.Arena
import com.dumbdogdiner.warrior.api.translation.Placeholders
import com.dumbdogdiner.warrior.util.DefaultMessages
import com.dumbdogdiner.warrior.Warrior
import com.dumbdogdiner.warrior.api.sessions.SessionType
import com.dumbdogdiner.warrior.api.sessions.ArenaSession
import com.dumbdogdiner.warrior.user.User
import com.dumbdogdiner.warrior.util.JSONUtil
import org.bukkit.entity.Player
import java.io.File
import java.util.ArrayList

class ArenaManager : WithWarriorPlugin, WarriorArenaManager {

    private val arenaMap = HashMap<Int, Arena>()
    override fun registerArena(arena: Arena) {
        arenaMap[arenaMap.size] = arena
    }

    override fun getArenas(): List<Arena> {
        return ArrayList(arenaMap.values)
    }

    override fun get(arenaName: String): Arena? {
        for (a in arenaMap.values) {
            if (a.name.equals(arenaName, ignoreCase = true)) return a
        }
        return null
    }

    override fun loadArenas() {
        val dataFolder = File(JSONUtil.ARENA_DATA_PATH)
        val files = dataFolder.listFiles() ?: return

        var i = 0
        for (f in files) {
            registerArena(Arena(f, i))
            i++
        }

        val msg = Placeholders.applyPlaceholders(DefaultMessages.LOADED_OBJECT, object : HashMap<String, String>() {
            init {
                put("amount", i.toString())
                put("type", "arena")
            }
        })
        getLogger().info(msg)
    }

    fun getPlayers(arena: Arena?): MutableList<User> {
        return Warrior.userCache.getListOf { user: User ->
            val session = user.session!!

            if(session is ArenaSession && session.type == SessionType.GAME)
                return@getListOf session.arena == arena

            false
        }.toMutableList()
    }

    override fun isPlaying(player: Player): Boolean {
        val user = Warrior.userCache[player.uniqueId]
        return if (user.session == null) false else user.session!!.type == SessionType.GAME
    }

    override fun getSession(player: Player): ArenaSession? {
        if (isPlaying(player)) {
            val user = Warrior.userCache[player.uniqueId]
            return user.session as ArenaSession
        }
        return null
    }

    override fun remove(a: Arena) {
        arenaMap.remove(a.id)
        JSONUtil.removeFile(a)
    }
}